ethexe/tx-pool/src/validation.rs

```diff
@@ -0,0 +1,255 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Transactions validation.
+
+use crate::SignedOffchainTransaction;
+use anyhow::{anyhow, Context, Result};
+use ethexe_db::Database;
+use ethexe_signer::ToDigest;
+use parity_scale_codec::Encode;
+
+// TODO #4424
+
+/// Main transaction pool tx validator.
+///
+/// Basically consumes a transaction and runs all the defined checks on it.
+/// The checks are defined through the `with_*_check` methods.
+pub(crate) struct TxValidator {
+    transaction: SignedOffchainTransaction,
+    db: Database,
+    signature_check: bool,
+    mortality_check: bool,
+    uniqueness_check: bool,
+}
+
+impl TxValidator {
+    pub(crate) fn new(transaction: SignedOffchainTransaction, db: Database) -> Self {
+        Self {
+            transaction,
+            db,
+            signature_check: false,
+            mortality_check: false,
+            uniqueness_check: false,
+        }
+    }
+
+    pub(crate) fn with_all_checks(self) -> Self {
+        self.with_signature_check()
+            .with_mortality_check()
+            .with_uniqueness_check()
+    }
+
+    pub(crate) fn with_signature_check(mut self) -> Self {
+        self.signature_check = true;
+        self
+    }
+
+    pub(crate) fn with_mortality_check(mut self) -> Self {
+        self.mortality_check = true;
+        self
+    }
+
+    pub(crate) fn with_uniqueness_check(mut self) -> Self {
+        self.uniqueness_check = true;
+        self
+    }
+}
+
+impl TxValidator {
+    /// Runs all stateful and stateless sync validators for the transaction.
+    pub(crate) fn validate(self) -> Result<SignedOffchainTransaction> {
+        if self.signature_check {
+            self.check_signature()?;
+        }
+
+        if self.mortality_check {
+            self.check_mortality()?;
+        }
+
+        if self.uniqueness_check {
+            self.check_uniqueness()?;
+        }
+
+        Ok(self.transaction)
+    }
+
+    /// Validates transaction signature.
+    fn check_signature(&self) -> Result<()> {
+        let tx_digest = self.transaction.encode().to_digest();
+        let signature = crate::tx_signature(&self.transaction)?;
+
+        signature.verify_with_public_key_recover(tx_digest)
+    }
+
+    /// Validates transaction mortality.
+    ///
+    /// Basically checks that transaction reference block hash is within the recent blocks window.
+    fn check_mortality(&self) -> Result<()> {
```

Comment:
for each check pls follow the same as for check_within_n_blocks: Result<bool>, err bailing only in case of storage invariants braking

Reply:
done